src/autocomplete/shared/extract-statement-positions-from-query.ts

```diff
@@ -0,0 +1,140 @@
+
+import {GetParseTree, LexerConstructor, ParserConstructor} from './autocomplete-types';
+import {createParser} from './query';
+import {SqlErrorListener} from './sql-error-listener';
+
+export interface StatementPosition {
+    startIndex: number;
+    endIndex: number;
+}
+
+export enum StatementExtractionStrategy {
+    Autocomplete = 'autocomplete',
+    Tokens = 'tokens',
+}
+
+export interface ExtractStatementPositionsResult {
+    statementPositions: StatementPosition[];
+    strategy: StatementExtractionStrategy;
+}
+
+export function extractStatementPositionsFromQuery<L extends LexerType, P extends ParserType>(
+    query: string,
+    Lexer: LexerConstructor<L>,
+    Parser: ParserConstructor<P>,
+    whitespaceToken: number,
+    emptySpaceTokens: number[],
+    endStatementToken: number,
+    statementRule: number,
+    getParseTree: GetParseTree<P>,
+): ExtractStatementPositionsResult {
+    const parser = createParser(Lexer, Parser, query);
+    const {tokenStream} = parser;
+    const errorListener = new SqlErrorListener(whitespaceToken);
+
+    parser.removeErrorListeners();
+    parser.addErrorListener(errorListener);
+    getParseTree(parser);
+
+    const autocompleteStatementPositions = parseAutocompleteStatements(
+        parser,
+        tokenStream,
+        statementRule,
+    );
+    if (autocompleteStatementPositions.length) {
+        return {
+            statementPositions: autocompleteStatementPositions,
+            strategy: StatementExtractionStrategy.Autocomplete,
+        };
+    }
+
+    let statementStartIndex = 0;
+    let processingNewStatement = false;
+
+    const statementPositions: StatementPosition[] = [];
+
+    for (let index = 0; index < tokenStream.size - 1; index++) {
+        const token = tokenStream.get(index);
+        const isEndStatementToken = token.type === endStatementToken;
+        const isEmptyToken = emptySpaceTokens.includes(token.type);
+
+        if (!processingNewStatement && isEmptyToken) {
+            continue;
+        }
+
+        if (!processingNewStatement) {
+            processingNewStatement = true;
+            statementStartIndex = token.start;
+        }
+
+        if (isEndStatementToken && statementStartIndex === token.start) {
+            processingNewStatement = false;
+            continue;
+        }
+
+        const isLastToken = index === tokenStream.size - 2;
+        if (isEndStatementToken || isLastToken) {
+            const tokenTextLength = token.text?.length || 0;
+            const statementEndIndex = token.start + tokenTextLength;
+
+            const statementAbsolutePosition: StatementPosition = {
+                startIndex: statementStartIndex,
+                endIndex: statementEndIndex,
+            };
+
+            statementPositions.push(statementAbsolutePosition);
+        }
+
+        if (isEndStatementToken) {
+            processingNewStatement = false;
```

Comment:
Let's move it to a separate function for better readability? This function feels bloated

Reply:
For better readability, you should just return statementPositions, without strategy. So that `extractStatementPositionsFromQuery` handles strategies.